Mathlib/Tactic/Linter/TextBased.lean

```diff
@@ -0,0 +1,203 @@
+  | StyleError.fileTooLong _ _ => StyleError.fileTooLong 0 0
+
+/-- Careful: we do not want to compare `ErrorContexts` exactly; we ignore some details. -/
+instance : BEq ErrorContext where
+  beq ctx ctx' :=
+      -- XXX: `lint-style.py` was calling `resolve()` on the path before before comparing them
+      -- should we also do so?
+      ctx.path == ctx'.path
+      -- We completely ignore line numbers of errors. Not sure if this is best.
+      -- We normalise errors before comparing them.
+      && (ctx.error).normalise == (ctx'.error).normalise
+
+/-- Output the formatted error message, containing its context. -/
+def outputMessage (errctx : ErrorContext) : String :=
+  -- We are outputting for github: duplicate file path, line number and error code,
+  -- so that they are also visible in the plain text output.
+  let path := errctx.path
+  let nr := errctx.lineNumber
+  let code := errctx.error.errorCode
+  s!"::ERR file={path},line={nr},code={code}::{path}:{nr} {code}: {errctx.error.errorMessage}"
+
+/-- Try parsing an `ErrorContext` from a string: return `some` if successful, `none` otherwise. -/
+def parse?_errorContext (line : String) : Option ErrorContext := Id.run do
+  -- Lines starting with "-- " are treated as a comment and ignored.
+  if line.startsWith "-- " then
+    return none
+  let parts := line.split (fun c ↦ c == ' ')
+  match parts with
+    | filename :: ":" :: "line" :: _line_number :: ":" :: error_code :: ":" :: error_message =>
+      -- Turn the filename into a path. XXX: is there a nicer way to do this?
+      -- Invariant: `style-exceptions.txt` always contains Unix paths
+      -- (because, for example, in practice it is updated by CI, which runs on unix).
+      -- Hence, splitting and joining on "/" is actually somewhat safe.
+      let path : System.FilePath := System.mkFilePath (filename.split (fun c ↦ c == '/'))
+      -- Parse the error kind from the error code, ugh.
+      -- NB: keep this in sync with `StyleError.errorCode` above!
+      let err : Option StyleError := match error_code with
+        -- Use default values for parameters which are normalised.
+        -- NB: keep this in sync with `normalise` above!
+        | "ERR_NUM_LIN" =>
+          -- Parse the error message in the script. `none` indicates invalid input.
+          match (error_message.get? 0, error_message.get? 3) with
+          | (some limit, some current) =>
+            match (String.toNat? limit, String.toNat? current) with
+            | (some size_limit, some current_size) =>
+              some (StyleError.fileTooLong current_size size_limit)
+            | _ => none
+          | _ => none
+        | _ => none
+      -- Omit the line number, as we don't use it anyway.
+      return (err.map fun e ↦ (ErrorContext.mk e 0 path))
+    -- XXX: print an error on any lines which don't match the particular format.
+    | _ => -- IO.println s!"Invalid input file: line {line} doesn't match any particular format"
```

Comment:
Printing an error here doesn't work because the monads don't work out. Quickly trying to fix that failed - a suggestion is welcome.

Reply:
Why not just return a fake `ErrorContext` here?

Reply:
Sure, I could to this - but I don't see the advantage over just returning `none` (as I'm doing now). Can you articulate why you consider that better?

Printing an error is merely for validating the style exceptions file; I'm also fine with leaving this out: the style exceptions file is usually automatically generated, after all! (And if not, we should make updating it easier, as #13623 does.)

Reply:
Okay. Do you want to just clean up the comments here, and explain the situation (maybe that it would be nice to report an error, but awkward, and hopefully because the file is automatically generated unnecessary)?

Reply:
Sounds good; I added such a comment.